### PR TITLE
Refactor existing tests

### PR DESCRIPTION
**Pull request by weienwong**
Some major changes made to the how testing is being done in the node-sdk package:

**Using the serial keyword**
One major change is adding the `serial` keyword. The reason for this change is because each test uses shared state ie. `t.context`. Without using the `serial` keyword we were experiencing unexpected failures relating to authentication. This issue is outlined in the doc: https://github.com/avajs/ava/blob/master/docs/08-common-pitfalls.md

This also resolves to auth issues that were being experienced during testing.

Also to clarify, when running tests using the `test()` statement, each test is run concurrently, not asynchronously. (https://github.com/avajs/ava/blob/master/docs/08-common-pitfalls.md)

**Additional testing**
This goal of this PR is to increase our test coverage as well as handle any known use cases. Test coverage for code in the node-sdk lies between 95%-100%

**Additional documentation**
The goal of this PR is to document existing tests.

** Testing **

The following code was not able to get tested for the following reasons:

`src/sdk.js`
- line 37 can’t test an if statement 
 ```
37 if (actions[funcName].bind) {
38   this.action[funcName] = actions[funcName].bind(this);
39 }
```

`src/services/service.js`
- line 120 - deprecated code
```
118    // DEPRECATED
119    if (params.usesXAuthHeader) {
120      opts.headers["X-Auth"] = this.token;
121    }
```

`src/account/account.js`
- line 10 - options param check
```
9    this.baseAPI =
10      options.accountsAPIURL ||
11      process.env.ZESTY_ACCOUNTS_API ||
12      `https://accounts.api.zesty.io/v1`;
```

`src/services/auth.js`
- line 9 - options param check
```
8    this.authURL =
9      options.authURL ||
10      process.env.ZESTY_AUTH_API ||
11      "https://auth.api.zesty.io";
```

`src/actions/find-and-replace/find-and-replace.js`
- line 22 cannot be tested because the function is unable to perform a search and replace on a single item without its content model ZUID. The function is dynamically called and parameters are not being passed in properly. To invoke `getItem`, it requires both a content model ZUID AND an item ZUID. **This functionality will need to be fixed.**
```
21  // Dynamically determine SDK call
22  const funcName = prefix == "6" ? "getItems" : "getItem";
23  const res = await this.instance[funcName](zuid);
```

- line 33-35 cannot be tested because they do not return a value and only have side effects
```
33      if (value) {
34        const newValue = value.replace(substr, newSubstr);
35        if (value !== newValue) {
36          item.data[fieldName] = newValue;
37          dirty.push(item);
38        }
39      }
```

`src/services/media/media.js`
- line 56: unable to set the site ID manually
```
55    if (!siteId) {
56      throw new Error("Failed to resolve instance ZUID");
57    }
```
- line 144-150 cannot be tested because they only have side effects 
```
143    if (opts.groupZUID) {
144      payload.append("group_id", opts.groupZUID);
145    } else {
146      payload.append("group_id", binZUID);
147    }
148
149    if (opts.userZUID) {
150      payload.append("user_id", opts.userZUID);
151    }
152    if (opts.title) {
153      payload.append("title", opts.title);
154    }
155    if (opts.fileName) {
156      payload.append("filename", opts.fileName);
157    }
```

`src/services/instance/items`
line 322: This is a rather complex test case which requires specific data values.  In order for line 322 to be executed, the following would need to happen.
- an item that contains content that MATCHES the `pathPart` value but its `pathPart` does not contain that value
- this upsert case is fairly niche
```
312        const res = await this.findItem(path);
313
314        if (Array.isArray(res.data) && res.data.length) {
315          const item = res.data.find(item => item.web.pathPart === path);
316          if (item) {
317            // Ensure required masterZUID is set for updates
318            payload.meta.masterZUID = item.meta.ZUID;
319
320            return await this.updateItem(modelZUID, item.meta.ZUID, payload);
321          } else {
322            return await this.createItem(modelZUID, payload);
323          }
324        } else {
325          return await this.createItem(modelZUID, payload);
326        }
```

**Test naming conventions**

I noticed tests have certain naming conventions such as “fetchModelFields:200” - while test names describe accurately what the expected test does, it only applies to a specific subset of tests. Our testing doesn’t just cover HTTP codes anymore but we also check for exceptions and error messages.

**SDK implementation changes**
`src/services/media/media.js`
- line 21 change ID to ZUID

`src/services/instance/views/index.js`
- remove duplicate definition of `getView()`

`src/services/instance/head-tags/index.js`
- line 29 change `fetchAuditLog` to `getHeadTag`

`src/services/instance/settings/index.js`
- line 6 change setting ID to setting ZUID

**API Issues**
I have also come across and documented an issue regarding creating stylesheet that contains code using the instances-api [here](https://github.com/zesty-io/instances-api/issues/271)

### CODE REVIEW

File: src/services/instance/items/items.test.js [hunk 1]
```diff
@@ -18,13 +18,13 @@ const TEST_ITEM_JSON = JSON.parse(
 
 test.beforeEach(authContext);
 
-test("fetchItems:200", async t => {
+test.serial("fetchItems:200", async t => {
```
Comment (protometheus) [inline src/services/instance/items/items.test.js:21]: comment in the README explaining why these are running serially. Include all of the other tests that are also now serially run.

File: src/services/account/account.test.js [hunk 2]
```diff
@@ -5,16 +5,21 @@ const test = require("ava");
...
 
-test("getInstanceUsers:200", async t => {
+test.serial("getInstanceUsers:200", async t => {
   const res = await t.context.sdk.account.getInstanceUsers();
   t.is(res.statusCode, 200);
   t.truthy(Array.isArray(res.data));
   t.truthy(res.data.length > 0);
 });
+
+test.serial("getSiteId:200", async t => {
```
Comment (protometheus) [inline src/services/account/account.test.js:22]: needs comment

File: src/sdk.test.js [hunk 3]
```diff
@@ -15,17 +15,28 @@ async function authedSDK() {
...
+test.serial("requires token", t => {
   try {
     new SDK(process.env.ZESTY_INSTANCE_ZUID);
   } catch (err) {
     t.is(err.message, "SDK:constructor() missing required `token` parameter");
   }
 });
 
-test("authenticated", async t => {
+test.serial("intialize invalid token", async t => {
```
Comment (protometheus) [inline src/sdk.test.js:26]: needs comment

File: src/services/instance/items/items.test.js (hunk 1 above)
Comment (protometheus) [reply src/services/instance/items/items.test.js:21]: also include the extra variables you introduced under the `Development` section

**Comment by weienwong** [comment]
@shrunyan - I have updated the PR description

**Comment by shrunyan** [comment]
> Without using the serial keyword we were experiencing unexpected failures relating to authentication. This issue is outlined in the doc: https://github.com/avajs/ava/blob/master/docs/08-common-pitfalls.md

The docs mention;
> Concurrent tests allow for asynchronous tests to execute more quickly, but if they rely on shared state you this may lead to unexpected test failures.

Please provide an explanation of _why/how_ using `t.context` is causing authentication to fail. I'm not aware of any shared state, as pointed out in the docs, between the tests which would cause authentication failure. 

---

> Also to clarify, when running tests using the test() statement, each test is run concurrently, not asynchronously. (https://github.com/avajs/ava/blob/master/docs/08-common-pitfalls.md)

Correct. The issue is serial vs concurrent. We want these to be ran concurrently, not in series, so we can have many many tests and the suite runs quickly.

**Comment by weienwong** [comment]
> Please provide an explanation of _why/how_ using `t.context` is causing authentication to fail. I'm not aware of any shared state, as pointed out in the docs, between the tests which would cause authentication failure.

Before each test is run, `beforeEach(authContext)` is being used to create an authenticated SDK context by logging in with a username and password. If we were to avoid using the `test.serial` statement, each test will have to authenticate concurrently before it performs a test. This may be what is causing failed login issues and 401, thus causing the tests to fail.

Using `test.serial()` to serialize each login and test, we prevent the above scenario of failed logins.

> Correct. The issue is serial vs concurrent. We want these to be ran concurrently, not in series, so we can have many many tests and the suite runs quickly.

Agreed, concurrent tests are faster. However the tests are less stable. It is possible to run these tests concurrently without failed logins, if we were to use dev tokens per request, eliminating the need to authenticate for every test. 

The dev token currently does not grant permissions to perform CRUD on media bins, files, etc. causing our media related tests to fail.

**Pull request closed without merging**